### PR TITLE
COCOS-214 - Improve manager resiliance by tracking vms on restart

### LINKED ISSUE #214: Feature: Make Manager robust and resilient
### Is your feature request related to a problem? Please describe.

Manager can die, and CVMs can be left hanging.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Manager must be restarted on failure, but it also must re-establish connections to deployed VMs. Agents in VMs must buffer (cache) logs during the `vsock` disconnections.

In the case that it cannot re-establish connections - maybe Manager needs to tear down VMs (i.e. stop them or kill them), in order to free the RAM. Appropriate messages then need to be sent to UI.

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by SammyOina**
# What type of PR is this?
This is a feature which persists vms on disk and allows manager to restart and track these vms

## What does this do?
- **New Features**
	- Introduced persistent state management for virtual machines, enabling saving, loading, and deleting VM states.
	- Added new methods for managing process identifiers within the virtual machine interface.

- **Bug Fixes**
	- Enhanced error handling in mock functions for better testing scenarios.

- **Chores**
	- Established a new directory for temporary data storage related to VM states.

## Which issue(s) does this PR fix/relate to?

- Related Issue #214
- Resolves #214

## Have you included tests for your changes?
Manually tested.

```shell
ps aux | grep qemu-system-x86_64                                                                 ✔ 
sammy     108667 10.8  5.9 3243188 1200784 pts/2 Sl   17:42   3:03 /usr/bin/qemu-system-x86_64 -enable-kvm -machine q35 -cpu EPYC -smp 4,maxcpus=64 -m 2048M,slots=5,maxmem=30G -drive if=pflash,format=raw,unit=0,file=/usr/share/edk2/x64/OVMF_CODE.fd,readonly=on -drive if=pflash,format=raw,unit=1,file=./img/OVMF_VARS.fd -netdev user,id=vmnic-2de3b110-e1c5-4df8-a3aa-551bba44e5ef,hostfwd=tcp::6100-:7002 -device virtio-net-pci,disable-legacy=on,iommu_platform=true,netdev=vmnic-2de3b110-e1c5-4df8-a3aa-551bba44e5ef,addr=0x2,romfile= -device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=4 -vnc :0 -kernel img/bzImage -append "quiet console=null rootfstype=ramfs" -initrd img/rootfs.cpio.gz -nographic -monitor pty
sammy     112831  0.0  0.0   9436  2036 pts/0    S+   18:10   0:00 grep qemu-system-x86_64

ps aux | grep qemu-system-x86_64                                                                 ✔ 
sammy     113971  0.0  0.0   9436  2156 pts/0    S+   18:18   0:00 grep qemu-system-x86_64
    ~/Documents/buildroot    master ⇣224  ls /tmp/cocos                                                                                    ✔ 
2.json


go run ./test/computations/main.go ./build/dist/lin_reg2 ../rsa_public.key false ./test/manual/data/iris.csv
{"time":"2024-08-27T18:16:38.89721241+03:00","level":"INFO","msg":"manager_test_server service gRPC server listening at :7001 without TLS"}
{"time":"2024-08-27T18:16:44.339990316+03:00","level":"DEBUG","msg":"received who am on ip address [::1]:54086"}
stopping
received agent event
&{event_type:"stop-computation"  timestamp:{seconds:1724771925  nanos:399632604}  computation_id:"1"  originator:"manager"  status:"complete"}
&{computation_id:"1"}


MANAGER_GRPC_URL=localhost:7001 MANAGER_LOG_LEVEL=debug MANAGER_QEMU_USE_SUDO=false  MANAGER_QEMU_ENABLE_SEV=false MANAGER_QEMU_SEV_CBITPOS=51 MANAGER_QEMU_OVMF_CODE_FILE=/usr/share/edk2/x64/OVMF_CODE.fd MANAGER_QEMU_OVMF_VARS_FILE=./img/OVMF_VARS.fd MANAGER_QEMU_ENABLE_SEV_SNP=false go run main.go
{"time":"2024-08-27T18:16:44.328872768+03:00","level":"INFO","msg":"-enable-kvm -machine q35 -cpu EPYC -smp 4,maxcpus=64 -m 2048M,slots=5,maxmem=30G -drive if=pflash,format=raw,unit=0,file=/usr/share/edk2/x64/OVMF_CODE.fd,readonly=on -drive if=pflash,format=raw,unit=1,file=./img/OVMF_VARS.fd -netdev user,id=vmnic,hostfwd=tcp::7020-:7002 -device virtio-net-pci,disable-legacy=on,iommu_platform=true,netdev=vmnic,addr=0x2,romfile= -device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=3 -vnc :0 -kernel img/bzImage -append \"quiet console=null rootfstype=ramfs\" -initrd img/rootfs.cpio.gz -nographic -monitor pty"}
{"time":"2024-08-27T18:16:44.339605659+03:00","level":"INFO","msg":"Successfully reattached to VM process","computation":"1","pid":108667}
{"time":"2024-08-27T18:16:44.339650748+03:00","level":"INFO","msg":"Restored VM state","id":"1","computationId":"1"}
{"time":"2024-08-27T18:16:44.33977245+03:00","level":"INFO","msg":"Successfully reattached to VM process","computation":"2","pid":113115}
{"time":"2024-08-27T18:16:44.33980721+03:00","level":"INFO","msg":"Restored VM state","id":"2","computationId":"2"}
{"time":"2024-08-27T18:18:45.399689942+03:00","level":"INFO","msg":"Method Stop for computation took 193.289µs to complete"}
2024/08/27 18:18:49 traces export: Post "http://localhost:4318/v1/traces": dial tcp [::1]:4318: connect: connection refused

```

## Did you document any new/modified feature?
No

### CODE REVIEW

File: manager/qemu/persistence.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func NewFilePersistence(dir string) (Persistence, error) {
+	if err := os.MkdirAll(dir, 0o755); err != nil {
+		return nil, err
+	}
+	return &FilePersistence{dir: dir}, nil
+}
+
+func (fp *FilePersistence) SaveVM(state VMState) error {
```
Comment (drasko) [inline manager/qemu/persistence.go:39]: What is this doing? Saving VM on disk?

Never leave RAM, only RAM is in TEE!

**Review by drasko — CHANGES_REQUESTED** [review]

File: manager/qemu/persistence.go (hunk 1 above)
Comment (SammyOina) [reply manager/qemu/persistence.go:39]: this is on the host(manager) not in TEE

**Review by drasko — CHANGES_REQUESTED** [review]

File: manager/qemu/persistence.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	defer fp.lock.Unlock()
+
+	files, err := os.ReadDir(fp.dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var states []VMState
+	for _, file := range files {
+		if filepath.Ext(file.Name()) != ".json" {
```
Comment (drasko) [inline manager/qemu/persistence.go:62]: Move thsi to string constant

File: manager/qemu/persistence.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+func (fp *FilePersistence) SaveVM(state VMState) error {
+	fp.lock.Lock()
+	defer fp.lock.Unlock()
+
+	data, err := json.Marshal(state)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(filepath.Join(fp.dir, state.ID+".json"), data, 0o644)
```
Comment (drasko) [inline manager/qemu/persistence.go:48]: string constant

**Review by drasko — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
Merged! Great improvement!

**Pull request merged**